### PR TITLE
ARROW-3539: [CI/Packaging] Update scripts to build against vendored jemalloc

### PR DESCRIPTION
**Pull request by kszucs**
Actually to not install jemalloc :)

### CODE REVIEW

**Review by wesm — APPROVED** [review]
+1, now I understand =) Thanks

**Comment by kszucs** [comment]
on bionic the build fails with (regardless of this change)

```
-- extracting...
     src='/build/apache-arrow-0.12.0~dev20181017/cpp/thirdparty/jemalloc/17c897976c60b0e6e4f4a365c751027244dada7a.tar.gz'
     dst='/build/apache-arrow-0.12.0~dev20181017/cpp_build/jemalloc_ep-prefix/src/jemalloc_ep'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
read(): Is a directory
make[5]: warning: -jN forced in submake: disabling jobserver mode.
make[5]: *** INTERNAL: readdir: Bad file descriptor.  Stop.
make[5]: Entering directory '/build/apache-arrow-0.12.0~dev20181017/cpp_build/jemalloc_ep-prefix/src/jemalloc_ep'
make[5]: Leaving directory '/build/apache-arrow-0.12.0~dev20181017/cpp_build/jemalloc_ep-prefix/src/jemalloc_ep'
```

**Comment by kszucs** [comment]
@wesm how can I disable parallel make for jemalloc_ep?

**Comment by wesm** [comment]
See https://github.com/apache/arrow/commit/1c85e41e3241450bc357e0e42709124af9f88803

**Comment by kszucs** [comment]
Thanks! If that resolves the buonic build should I introduce a cmake flag for that?

**Comment by kszucs** [comment]
However that's another issue. I'm waiting for crossbow builds: https://github.com/kszucs/crossbow/branches/all?utf8=%E2%9C%93&query=build-335

**Comment by wesm** [comment]
Hm, do we understand the root cause of the parallel build failure? That seems odd

**Comment by kszucs** [comment]
Not yet, but there was a similar issue https://github.com/apache/arrow/pull/2245

**Comment by kou** [comment]
Using `make -j` in sub-`make` is fragile. And it's useless when we use `-j` in top-level `make`. So we should avoid it.

How about the following? We can use parallel build when we use Ninja CMake generator.

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 15ff5ec8..7540e782 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -309,6 +309,12 @@ if ("${MAKE}" STREQUAL "")
     endif()
 endif()
 
+if (${CMAKE_GENERATOR} MATCHES "Unix Makefiles")
+  set(MAKE_BUILD_ARGS "")
+else()
+  set(MAKE_BUILD_ARGS "-j")
+endif()
+
 # ----------------------------------------------------------------------
 # Find pthreads
 
@@ -736,9 +742,9 @@ if (ARROW_JEMALLOC)
     CONFIGURE_COMMAND ./autogen.sh "--prefix=${JEMALLOC_PREFIX}" "--with-jemalloc-prefix=je_arrow_" "--with-private-namespace=je_arrow_private_" "--disable-tls"
     ${EP_LOG_OPTIONS}
     BUILD_IN_SOURCE 1
-    BUILD_COMMAND ${MAKE} -j
+    BUILD_COMMAND ${MAKE} ${MAKE_BUILD_ARGS}
     BUILD_BYPRODUCTS "${JEMALLOC_STATIC_LIB}" "${JEMALLOC_SHARED_LIB}"
-    INSTALL_COMMAND ${MAKE} -j1 install)
+    INSTALL_COMMAND ${MAKE} install)
 
   # Don't use the include directory directly so that we can point to a path
   # that is unique to our codebase.
```

**Comment by kou** [comment]
`ninja`? `ninja` does parallel builds by default.

**Comment by kszucs** [comment]
Yes ninja :) Then do We need `set(MAKE_BUILD_ARGS "-j")`?

**Comment by kou** [comment]
> Then do We need `set(MAKE_BUILD_ARGS "-j")`?

Yes. Ninja users like fast build. So we should use `-j` for Ninja users. The patch enables `-j` all users except Make users.

**Comment by kou** [comment]
Yes. `ninja` requires the maximum number of jobs.

But `make` doesn't require the maximum number of jobs. If we omit the maximum number of jobs, `make` doesn't limit the concurrent number of jobs. It causes a problem when we have many source codes. If we process many source codes at once, system resources (normally, CPU and I/O) are conflicted. It reduces build performance. jemalloc only have 30 .c files. It's not very small but it will not use all system resources. It may be better that we use `-j4` or `-j8` for safety.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -309,6 +309,15 @@ if ("${MAKE}" STREQUAL "")
     endif()
 endif()
 
+# Using make -j in sub-make is fragile
+# see discussion https://github.com/apache/arrow/pull/2779
+if (${CMAKE_GENERATOR} MATCHES "Makefiles")
+    set(MAKE_BUILD_ARGS "")
+else()
+    # limit the maximum number of jobs for ninja
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: You may misunderstand. This argument is used by `make` not `ninja`.

**Comment by kou** [comment]
The CI failure is unrelated.
I restarted the failed task.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (kszucs) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: @kou No, I don't fully understand :)
My conclusion from your reasoning and patch was to not parallelize `make` at all, and parallelize but limit the number of jobs for `ninja -j4`. 

BTW The build is passing with the current changeset.

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Right.

**Review by kou — APPROVED** [review]
+1
I'll merge this.

**Comment by kou** [comment]
@kszucs This still have WIP label. Do you have more works for this?

**Comment by kszucs** [comment]
Removed it

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**